### PR TITLE
rt refine: prove decode priority corres

### PR DESCRIPTION
**Pull request by pierzchalski**
Signed-off-by: Edward Pierzchalski <ed.pierzchalski@data61.csiro.au>

### CODE REVIEW

**Review by MitchellBuckley — APPROVED** [review]
Love it!

**Review by corlewis — APPROVED** [review]
Nice! I'm also looking forward to that great spec cleanup in the wip commit.

**Comment by pierzchalski** [comment]
(There's a work-in-progress commit that shouldn't be on this branch, I'll clean it up later)

File: spec/abstract/Decode_A.thy [hunk 1]
```diff
@@ -320,26 +320,26 @@ definition
...
+             sc_ptr' \<leftarrow> liftE $ get_tcb_obj_ref tcb_sched_context tcb_ptr;
+             whenE (sc_ptr' \<noteq> None) $ throwError IllegalOperation;
+             sc \<leftarrow> liftE $ get_sched_context sc_ptr;
+             whenE (sc_tcb sc \<noteq> None) $ throwError IllegalOperation;
+             st \<leftarrow> liftE $ get_thread_state tcb_ptr;
+             released \<leftarrow> liftE $ get_sc_released sc_ptr;
+             whenE (ipc_queued_thread_state st \<and> \<not>released) $ throwError IllegalOperation;
+             returnOk $ ThreadControlSched tcb_ptr slot None None None (Some (Some sc_ptr))
+         odE
+     | _ \<Rightarrow> throwError (InvalidCapability 2)"
```
Comment (corlewis) [inline spec/abstract/Decode_A.thy:342]: I know that it's not actually part of this PR, but is it intentional that this has changed from `InvalidCapability 0` to `InvalidCapability 2`?

File: proof/refine/ARM/Tcb_R.thy [hunk 2]
```diff
@@ -2723,20 +2723,120 @@ lemma decodeSetSchedParams_wf[wp]:
...
+          apply (rule corres_guard_imp)
+            apply (case_tac "sc_cap = cap.NullCap"
+                   ; clarsimp simp: decode_update_sc_def)
+             apply (rule corres_split_eqrE)
+                apply (rule whenE_throwError_corres'; simp)
+                apply (rule corres_returnOkTT; simp)
+               apply simp
+               apply (rule gct_corres)
+              apply wpsimp+
+find_theorems "corres_underlying" "(_ \<Longrightarrow> _) \<Longrightarrow> _"
```
Comment (mktnk3) [inline proof/refine/ARM/Tcb_R.thy:2800]: Looks like this part of proof needs a lot of cleaning up.

File: proof/refine/ARM/Tcb_R.thy [hunk 3]
```diff
@@ -2723,20 +2723,120 @@ lemma decodeSetSchedParams_wf[wp]:
...
 
+(* FIXME RT: move to...? *)
+lemma cap_rel_valid_fh:
+  "cap_relation a b \<Longrightarrow> valid_fault_handler a = isValidFaultHandler b"
+  apply (case_tac a
+         ; case_tac b
+         ; simp add: valid_fault_handler_def isValidFaultHandler_def)
+  apply (rule iffI
+         ; clarsimp simp: has_handler_rights_def split: bool.split_asm)
+  done
```
Comment (mktnk3) [inline proof/refine/ARM/Tcb_R.thy:2734]: This can go in `StateRelation.thy`, I think.

File: proof/refine/ARM/Tcb_R.thy [hunk 4]
```diff
@@ -2723,20 +2723,120 @@ lemma decodeSetSchedParams_wf[wp]:
...
+                                      odE
+                                  | capability.SchedContextCap scPtr x \<Rightarrow> doE tcbSc <- withoutFailure $ threadGet tcbSchedContext tcb_ptr;
+                                                             whenE (tcbSc \<noteq> Nothing) $ throw Fault_H.syscall_error.IllegalOperation;
+                                                             scTcb <- withoutFailure $ mapScPtr scPtr scTCB;
+                                                             whenE (scTcb \<noteq> Nothing) $ throw Fault_H.syscall_error.IllegalOperation;
+                                                             blockedAndUnreleased <- withoutFailure $ isBlocked tcb_ptr `~andM~` (liftM Not $ scReleased scPtr);
+                                                             whenE blockedAndUnreleased $ throw Fault_H.syscall_error.IllegalOperation;
+                                                             returnOk $ Just scPtr
+                                                         odE
+                                  | _ \<Rightarrow> throw $ Fault_H.syscall_error.InvalidCapability 2)")
```
Comment (mktnk3) [inline proof/refine/ARM/Tcb_R.thy:2790]: This is a big `prop_tac`. It probably makes sense to have this as a separate lemma? Or any reason you choose not to?

**Comment by pierzchalski** [comment]
@mktnk3 Sorry that I didn't sort this out in time, but there was a stray WIP commit on this PR for a while and it looks like your comments were about the contents of that commit. How does the (much more straightforward!) _intended_ PR look?

**Comment by mktnk3** [comment]
> @mktnk3 Sorry that I didn't sort this out in time, but there was a stray WIP commit on this PR for a while and it looks like your comments were about the contents of that commit. How does the (much more straightforward!) _intended_ PR look?

Right, yes, looks good to me! Ready to merge now?

**Review by mktnk3 — APPROVED** [review]

**Comment by pierzchalski** [comment]
[post-rebase testboard](https://bamboo.ts.data61.csiro.au/browse/L4V-TESTBOARD-656)

**Pull request merged**